File: cs/Problems/Day11.cs
namespace AOC2023.Problems;

public class Day11 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => SumDistanceBetweenGalaxies(input);

    private int SumDistanceBetweenGalaxies(ReadOnlySpan<char> input)
    {
        Span<Range> lineRanges = stackalloc Range[140];
        int lines = input.Split(lineRanges, InputReader.NewLine);

        var emptyCols = new List<int>(lines);
        var emptyRows = new List<int>(lines);

        for (int j = 0; j < lines; j++)
        {
            emptyCols.Add(lines - j);
            emptyRows.Add(lines - j);
        }

        Span<Point> galaxies = stackalloc Point[450];
        int galaxyCount = 0;

        for (int j = lines - 1; j >= 0; j--)
        {
            var line = input[lineRanges[j]];
            for (int i = line.Length - 1; i >= 0; i--)
            {
                if (line[i] == '#')
                {
                    galaxies[galaxyCount++] = new Point(i, j);
                    emptyCols.Remove(j);
                    emptyRows.Remove(i);
                }
            }
        }

        for (int i = 0; i < galaxyCount; i++) // Universe expansion.
        {
            var galaxy = galaxies[i];

            foreach (int col in emptyCols)
                galaxy.Y += col <= galaxy.Y ? 1 : 0;

            foreach (int row in emptyRows)
                galaxy.X += row <= galaxy.X ? 1 : 0;

            galaxies[i] = galaxy;
        }

        int sumOfDistances = 0;

        for (int i = 0; i < galaxyCount - 1; i++)
        {
            for (int j = i + 1; j < galaxyCount; j++)
            {
                int distance = Math.Abs(galaxies[j].X - galaxies[i].X) + Math.Abs(galaxies[j].Y - galaxies[i].Y);
                sumOfDistances += distance;
            }
        }

        return sumOfDistances;
    }

    record struct Point(int X, int Y);
}

File: go/problems/day11.go
package problems

import (
	"image"
	"slices"
	"strings"
)

type Day11 struct{}

func (d Day11) Solve(input string) int64 {
	return sumDistanceBetweenGalaxies(input)
}

func sumDistanceBetweenGalaxies(input string) int64 {
	lines := strings.Split(input, NewLine)
	l := len(lines)

	emptyCols := make([]int, l)
	emptyRows := make([]int, l)
	galaxies := make([]*image.Point, 0)

	for i := range lines {
		emptyCols[i] = l - i
		emptyRows[i] = l - i
	}

	for y, line := range lines {
		for x, char := range line {
			if char == '#' {
				galaxies = append(galaxies, &image.Point{X: x, Y: y})
				emptyCols = removeElement(emptyCols, y)
				emptyRows = removeElement(emptyRows, x)
			}
		}
	}

	for _, galaxy := range galaxies { // Universe expansion.
		for _, col := range emptyCols {
			if col <= galaxy.Y {
				galaxy.Y++
			}
		}
		for _, row := range emptyRows {
			if row <= galaxy.X {
				galaxy.X++
			}
		}
	}

	var totalDist int64 = 0

	for i := 0; i < len(galaxies)-1; i++ {
		for j := i + 1; j < len(galaxies); j++ {
			var dist int64 = int64(abs(galaxies[j].X-galaxies[i].X) + abs(galaxies[j].Y-galaxies[i].Y))
			totalDist += dist
		}
	}

	return totalDist
}

func abs(x int) int {
	if x < 0 {
		return x * -1
	}
	return x
}

func removeElement(arr []int, el int) []int {
	if inx := slices.Index(arr, el); inx >= 0 {
		return slices.Delete(arr, inx, inx+1)
	}
	return arr
}
